compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/ReshapeFusion.cpp

```diff
@@ -652,6 +652,93 @@ struct DropScatterUnitIndexDepth final : public OpRewritePattern<ScatterOp> {
+      return rewriter.notifyMatchFailure(
+          gatherOp, "dropping unit dims from batch rank 1 is not supported.");
+    }
+
+    Location loc = gatherOp.getLoc();
+
+    auto rankReduceOperand = [&rewriter,
+                              &loc](int64_t numDims, Value operand,
+                                    ShapedType ty) -> FailureOr<Value> {
+      ArrayRef<int64_t> shape = ty.getShape();
+
+      // Find list of dims to drop and the target shape.
+      SmallVector<bool> unitDims(shape.size(), false);
+      for (auto [isUnitDim, size] : llvm::zip_equal(unitDims, shape)) {
+        if (size == 1) {
+          isUnitDim = true;
+        }
+      }
+
+      // Because gather/scatter like to define special behavior to allow eliding
+      // the coordinate dimension, we have to make sure we don't generate a 0-D
+      // tensor.
+      auto slice = MutableArrayRef(unitDims).take_front(numDims);
+      if (llvm::all_of(slice, [](bool x) { return x; })) {
+        slice.back() = false;
+      }
+
+      SmallVector<int64_t> targetShape;
+      for (int i = 0; i < numDims; ++i) {
+        if (!unitDims[i]) {
+          targetShape.push_back(shape[i]);
+        }
+      }
+
+      ArrayRef<int64_t> remaining = shape.drop_front(numDims);
+      targetShape.append(remaining.begin(), remaining.end());
+
+      // No unit dims dropped.
+      if (targetShape.size() == shape.size()) {
+        return failure();
+      }
+
+      // Drop unit dims using extract_slice.
+      FailureOr<Value> rankReducingExtract =
+          tensor::ExtractSliceOp::rankReduceIfNeeded(rewriter, loc, operand,
+                                                     targetShape);
+      assert(succeeded(rankReducingExtract) && "not a unit-extent collapse");
+      return rankReducingExtract.value();
+    };
+
+    FailureOr<Value> newIndices =
+        rankReduceOperand(gatherOp.getBatchRank(), gatherOp.getIndices(),
+                          gatherOp.getIndicesType());
+    FailureOr<Value> newOutput =
+        rankReduceOperand(gatherOp.getBatchRank(), gatherOp.getOutput(),
+                          gatherOp.getOutputType());
+
+    if (failed(newIndices) || failed(newOutput)) {
+      return failure();
+    }
+
+    // New Gather.
+    auto newGather = rewriter.create<GatherOp>(
+        gatherOp.getLoc(), TypeRange{newOutput->getType()},
+        ValueRange{gatherOp.getSource(), newIndices.value()},
+        ValueRange{newOutput.value()}, gatherOp.getDimensionMap());
+
+    Value dest = gatherOp.getOutput();
+    int64_t rank = gatherOp.getOutputType().getRank();
+    SmallVector<OpFoldResult> offsets(rank, rewriter.getIndexAttr(0));
+    SmallVector<OpFoldResult> sizes =
+        tensor::getMixedSizes(rewriter, loc, dest);
+    SmallVector<OpFoldResult> strides(rank, rewriter.getIndexAttr(1));
+
+    rewriter.replaceOpWithNewOp<tensor::InsertSliceOp>(
+        gatherOp, newGather.getResult(0), dest, offsets, sizes, strides);
```

Comment:
I'd remove the `// New Gather.` comment because it does not provide values to me. The code is already self describing, because you name it `newGather`.

optional nit: I'd remove all the blank lines because they belong to the same chunk to me.